### PR TITLE
8282578: AIOOBE in javax.sound.sampled.Clip

### PR DESCRIPTION
**Pull request by azuev-java**
Add try/catch clause to ignore an exception since it is harmless for we isolated
the massge data before passing it ro processor.
Add test case.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8282578](https://bugs.openjdk.org/browse/JDK-8282578): AIOOBE in javax.sound.sampled.Clip

### Reviewers
 * [Phil Race](https://openjdk.org/census#prr) (@prrace - **Reviewer**) ⚠️ Review applies to [d085cfb1](https://git.openjdk.org/jdk/pull/9016/files/d085cfb10f6a1203e65b7ba2e98c20221fe41f7d)
 * [Alexey Ivanov](https://openjdk.org/census#aivanov) (@aivanov-jdk - **Reviewer**) ⚠️ Review applies to [d085cfb1](https://git.openjdk.org/jdk/pull/9016/files/d085cfb10f6a1203e65b7ba2e98c20221fe41f7d)
 * [Alexander Zvegintsev](https://openjdk.org/census#azvegint) (@azvegint - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/9016/head:pull/9016` \
`$ git checkout pull/9016`

Update a local copy of the PR: \
`$ git checkout pull/9016` \
`$ git pull https://git.openjdk.org/jdk pull/9016/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 9016`

View PR using the GUI difftool: \
`$ git pr show -t 9016`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/9016.diff">https://git.openjdk.org/jdk/pull/9016.diff</a>

</details>

### CODE REVIEW

**Comment by victordyakov** [comment]
@azvegint @prrace please review

**Comment by mrserb** [comment]
Probably it is better to check the data length for each sys message and discard it if the data is too small? Ignoring all possible AIOOBE  from this large method which calls many other large methods from SoftVoice/SoftTuning may hide some other bugs.

**Review by prrace — APPROVED** [review]

**Comment by azuev-java** [comment]
> Probably it is better to check the data length for each sys message and discard it if the data is too small? Ignoring all possible AIOOBE from this large method which calls many other large methods from SoftVoice/SoftTuning may hide some other bugs.

No, because determining the correct length of the message will require basically to parse it all, the correct length could be  a 2 bytes or hundreds of bytes - in order to determine we have to process the message completely.

**Comment by mrserb** [comment]
> No, because determining the correct length of the message will require basically to parse it all, the correct length could be a 2 bytes or hundreds of bytes - in order to determine we have to process the message completely.

There are 64 usages of `data` in that method most of them use constants like 1/2/3 and other usages are in the loops, both can be easily checked. There is also code like ` int[] destinations = new int[(data.length - 7) / 2];` which could cause the NegativeArrayException, or probably some others. It s better to check the out of bands access before access than to have a try/catch block for 300 lines of code.

**Review by kristylee88 — APPROVED** [review]

**Review by aivanov-jdk — APPROVED** [review]

**Comment by AJ1032480** [comment]
pre submit tests failed*

**Comment by azuev-java** [comment]
Just uploaded the new version with proper length analysis so would be thankful for the continuation of the review.

**Comment by azuev-java** [comment]
@azvegint Can you please take a look?

File: src/java.desktop/share/classes/com/sun/media/sound/SoftTuning.java [hunk 1]
```diff
@@ -89,17 +89,31 @@ private boolean checksumOK2(byte[] data) {
...
+                    break;
+                }
                 int subid2 = data[4] & 0xFF;
                 switch (subid2) {
                 case 0x01: // BULK TUNING DUMP (NON-REAL-TIME)
                 {
                     // http://www.midi.org/about-midi/tuning.shtml
                     //if (!checksumOK2(data))
                     //    break;
+                    if (data.length < 406) {
```
[the remark this answers is not in the document]
Comment (azuev-java) [reply src/java.desktop/share/classes/com/sun/media/sound/SoftTuning.java:114]: Makes sense, fixed.

File: src/java.desktop/share/classes/com/sun/media/sound/SoftTuning.java [hunk 2]
```diff
@@ -115,7 +129,13 @@ public void load(byte[] data) {
                 case 0x02: // SINGLE NOTE TUNING CHANGE (REAL-TIME)
                 {
                     // http://www.midi.org/about-midi/tuning.shtml
+                    if (data.length < 7) {
+                        break;
+                    }
                     int ll = data[6] & 0xFF;
+                    if (data.length < ll * 4 + 8) {
```
[the remark this answers is not in the document]
Comment (azuev-java) [reply src/java.desktop/share/classes/com/sun/media/sound/SoftTuning.java:136]: Yes.

File: src/java.desktop/share/classes/com/sun/media/sound/SoftTuning.java [hunk 3]
```diff
@@ -193,6 +228,9 @@ public void load(byte[] data) {
                            // Real-Time/REAL-TIME)
                 {
                     // http://www.midi.org/about-midi/tuning-scale.shtml
+                    if (data.length < 21) {
```
[the remark this answers is not in the document]
Comment (azuev-java) [reply src/java.desktop/share/classes/com/sun/media/sound/SoftTuning.java:231]: Yes. Math is hard.

File: src/java.desktop/share/classes/com/sun/media/sound/SoftMainMixer.java [hunk 4]
```diff
@@ -123,304 +123,368 @@ public double[] get(int instance, String name) {
...
+                                {
+                                    if (data.length < 7) {
                                         break;
-                                    default:
+                                    }
+                                    int[] destinations = new int[(data.length - 7) / 2];
+                                    int[] ranges = new int[(data.length - 7) / 2];
+                                    int ix = 0;
+                                    for (int j = 6; j < data.length - 1; j += 2) {
+                                        destinations[ix] = data[j] & 0xFF;
```
[the remark this answers is not in the document]
Comment (azuev-java) [reply src/java.desktop/share/classes/com/sun/media/sound/SoftMainMixer.java:406]: Ok, i think there are 3 places total with this possibility when increment goes by 2 so fixed them all.

File: src/java.desktop/share/classes/com/sun/media/sound/SoftMainMixer.java [hunk 5]
```diff
@@ -123,304 +123,368 @@ public double[] get(int instance, String name) {
...
+                                case 0x06:  // SCALE/OCTAVE TUNING DUMP, 2 byte format
+                                case 0x07:  // SINGLE NOTE TUNING CHANGE (NON REAL-TIME)
+                                    // (BANK)
+                                {
+                                    if (data.length < 7) {
                                         break;
                                     }
-                                    default:
-                                        break;
+                                    // http://www.midi.org/about-midi/tuning_extens.shtml
```
[the remark this answers is not in the document]
Comment (azuev-java) [reply src/java.desktop/share/classes/com/sun/media/sound/SoftMainMixer.java:166]: Yes. That's the problem referencing the external site for information. Unfortunately the documentation layout on midi site changed and one needs to read trough all of it to figure out the correct path that both relates to the midi-1.0 standard that we support and is not overwritten by some later addendum making it non-informative.

File: src/java.desktop/share/classes/com/sun/media/sound/SoftMainMixer.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (azvegint) [reply src/java.desktop/share/classes/com/sun/media/sound/SoftMainMixer.java:406]: Length check won't help here:

```java
  int[] data = new int[100];
  if (data.length < 8) {
      return;
  }
  int[] destinations = new int[(data.length - 7) / 2];
  int[] ranges = new int[(data.length - 7) / 2];
  int ix = 0;
  for (int j = 6; j < data.length - 1; j += 2) {
      destinations[ix] = data[j] & 0xFF;
      ranges[ix] = data[j + 1] & 0xFF;
      ix++;
  }
```
`Exception in thread "main" java.lang.ArrayIndexOutOfBoundsException: Index 46 out of bounds for length 46`


We might want to add more test cases to the test.

File: src/java.desktop/share/classes/com/sun/media/sound/SoftMainMixer.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (azvegint) [reply src/java.desktop/share/classes/com/sun/media/sound/SoftMainMixer.java:166]: We might want to track this under some issue, if not done already.

[the remark this answers is not in the document]
Comment (azuev-java) [reply src/java.desktop/share/classes/com/sun/media/sound/SoftMainMixer.java:166]: Created JDK-8298319 to track this issue.

File: src/java.desktop/share/classes/com/sun/media/sound/SoftMainMixer.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (azuev-java) [reply src/java.desktop/share/classes/com/sun/media/sound/SoftMainMixer.java:406]: > We might want to add more test cases to the test.

That would be an extensive task, there is a set of the tests being run by third party - and that's them who created the initial issue. I do not think it is practical to bring all the cases they test here. That testing takes a very long time.

**Review by azvegint — APPROVED** [review]

**Comment by mrserb** [comment]
Thank you for removing that large try/catch block. Looks much better.

**Comment by aivanov-jdk** [comment]
> Thank you for removing that large try/catch block. Looks much better.

I agree, the code is clearer now.

I started reviewing the updated code but I haven't finished it. It's _integrated_ now.

**Pull request closed without merging**